File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第5章跳蚱蜢.cpp
//跳蚱蜢  lanqiaoOJ 642

#include<bits/stdc++.h>
using namespace std;
struct node{
	node(){}
	node(string ss, int tt){s = ss, t = tt;}
	string s;
	int t;
};
int cnt = 0; //统计计算了多少次
map<string, bool> mp;
queue<node> q;
void solve(){
	while(!q.empty()){
		node now = q.front();
		q.pop();
		string s = now.s;
		int step = now.t;
		if(s == "087654321"){        //到目标了
			cout << step << endl;    //输出跳跃步数
			cout << cnt << endl;     //计算了 1451452 次
			break;
		}
		int i;
		for(i = 0 ; i < 10 ; i++)    //找到盘子的位置
			if(s[i] == '0')
                break;
		for(int j = i - 2 ; j <= i + 2 ; j++){  //4种跳法
			int k = (j + 9) % 9;
			if(k == i)	continue;               //这是当前状态，不用检查
			string news = s;
			char tmp = news[i];
            news[i] = news[k];
            news[k] = tmp;  //跳到一种情况
            cnt ++;
			if(!mp[news]){         //判重：这个情况没有出现过
				mp[news] = true;
				q.push(node(news, step + 1));
			}
		}
	}
}
int main(){
	string s = "012345678";
	q.push(node(s, 0));
	mp[s] = true;
	solve();
	return 0;
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第5章跳蚱蜢.java
//跳蚱蜢 lanqiaoOJ 642


//(1)set
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set; 
public class Main { 
    static String target = "087654321"; 
    public static void main(String[] args) {
        bfs();
    } 
    static Set<String> set = new HashSet<String>();
    static int[] dir = new int[] { -2, -1, 1, 2 }; 
    public static void bfs() {
        Queue<String> que = new LinkedList<String>();
        que.add("012345678");
        set.add("012345678"); 
        int ans = 0;
        while (!que.isEmpty()) {
            int size = que.size();
            for (int k = 0; k < size; ++k) {
                String p = que.poll();
                if(p.equals(target)) {
                    System.out.println(ans);
                    return ;
                }
                char[] cp = p.toCharArray();
                int ind = 0;
                for (int i = 0; i < cp.length; ++i) {
                    if (cp[i] == '0') {
                        ind = i;
                        break;
                    }
                } 
                for (int i = 0; i < dir.length; ++i) {
                    int nind = (ind + dir[i] + 9) % 9;
                    char[] temcp = cp.clone();
                    temcp[ind] = cp[nind];
                    temcp[nind] = cp[ind];
                    String strp = new String(temcp);
                    if (!set.contains(strp)) {
                        que.add(strp);
                        set.add(strp);
                    }
                }
            }
            ++ ans;
        } 
    } 
}


//(2)map
/*
import java.util.ArrayDeque;
import java.util.HashMap;
class node{
    String string;
    int pos;
    int step;
}
public class Main {
    static ArrayDeque<node> queue = new ArrayDeque();
    static HashMap<String, Integer> map = new HashMap<String, Integer>();
    static String start = "012345678";
    static String end = "087654321";
    static int[] d = {1, 2, -1, -2};
    public static void main(String[] args) {
        bfs();
    }
    public static String swap(String str, int i, int j) {
        char[] c = str.toCharArray();
        char ch = c[i];
        c[i] = c[j];
        c[j] = ch;
        return new String(c);
    }
    public static void bfs() {
        node p = new node();
        p.string = start;
        p.pos = start.indexOf("0");
        p.step = 0;
        queue.offer(p);
        map.put(start, 1);
        while(!queue.isEmpty()) {
            node pp = new node();
            pp = queue.pop();

            if(pp.string.equals(end)) {
                System.out.println(pp.step);
                break;
            }
            for(int i = 0; i < 4; i++) {
                int pos = pp.pos + d[i];
                if(pos < 0)
                    pos += 9;
                if(pos > 8)
                    pos -= 9;

                String s = pp.string;
                String ss = swap(s, pp.pos, pos);
                if(!map.containsKey(ss)) {
                    node ppp = new node();
                    ppp.string = ss;
                    ppp.pos = pos;
                    ppp.step = pp.step + 1;
                    queue.offer(ppp);
                    map.put(ss, 1);
                }
            }
        }
    }
}

*/


